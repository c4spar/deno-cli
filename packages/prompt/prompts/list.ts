import { encode } from 'https://deno.land/std@v0.41.0/encoding/utf8.ts';
import { blue, bold, dim, underline, yellow } from 'https://deno.land/std@v0.41.0/fmt/colors.ts';
import { stripeColors } from '../../table/lib/utils.ts';
import { Figures } from '../lib/figures.ts';
import { GenericInput, GenericInputPromptOptions, GenericInputPromptSettings } from './generic-input.ts';

export interface ListPromptOptions extends GenericInputPromptOptions<string[]> {
    separator?: string;
}

export interface ListPromptSettings extends GenericInputPromptSettings<string[]> {
    separator: string;
}

export class List extends GenericInput<string[], ListPromptSettings> {

    public static async prompt( options: ListPromptOptions ): Promise<string[] | undefined> {

        return new this( {
            pointer: blue( Figures.POINTER_SMALL ),
            separator: ',',
            ...options
        } ).run();
    }

    protected getMessage(): string {

        let message = ` ${ yellow( '?' ) } ${ bold( this.settings.message ) }`;

        if ( typeof this.settings.default !== 'undefined' ) {
            message += dim( ` (${ this.settings.default.join( `${ this.settings.separator } ` ) })` );
        }

        return message;
    }

    protected setMessage( message: string ) {

        message += ' ' + this.settings.pointer + ' ';

        const length = encode( stripeColors( message ) ).length;

        const oldInput: string = this.input;
        const oldInputParts: string[] = oldInput.trimLeft().split( this.regexp() );

        this.input = oldInputParts.join( `${ this.settings.separator } ` );

        message += oldInputParts.map( ( val: string ) => underline( val ) ).join( `${ this.settings.separator } ` );

        const inputDiff = oldInput.length - this.input.length;

        this.index -= inputDiff;

        this.write( message );

        this.screen.cursorTo( length - 1 + this.index );
    }

    protected regexp() {
        return new RegExp( this.settings.separator === ' ' ? ` +` : ` *${ this.settings.separator } *` );
    }

    protected addChar( char: string ): void {
        switch ( char ) {
            case this.settings.separator:
                if ( this.input.length && this.input.trim().slice( -1 ) !== this.settings.separator ) {
                    super.addChar( char );
                }
                break;
            default:
                super.addChar( char );
        }
    }

    protected deleteChar(): void {
        if ( this.input[ this.index - 1 ] === ' ' ) {
            super.deleteChar();
        }
        super.deleteChar();
    }

    protected sanitize( value: string ): string[] {
        return value.trim().split( this.regexp() );
    }

    protected validate( value: string[] | undefined ): boolean | string {
        return Array.isArray( value ) && ( value.every( ( val: string ) => val.trim().length > 0 ) || 'Value cannot be empty.' );
    }

    protected transform( value: string[] ): string {
        return value.join( `${ this.settings.separator } ` );
    }
}
