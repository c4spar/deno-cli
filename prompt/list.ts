import { stripeColors } from "../table/utils.ts";
import { blue, underline } from "./deps.ts";
import { Figures } from "./figures.ts";
import {
  GenericInput,
  GenericInputKeys,
  GenericInputPromptOptions,
  GenericInputPromptSettings,
} from "./_generic-input.ts";

export type ListKeys = GenericInputKeys;

export interface ListOptions extends GenericInputPromptOptions<string[]> {
  separator?: string;
  minLength?: number;
  maxLength?: number;
  minTags?: number;
  maxTags?: number;
  keys?: ListKeys;
}

interface ListSettings extends GenericInputPromptSettings<string[]> {
  separator: string;
  minLength: number;
  maxLength: number;
  minTags: number;
  maxTags: number;
  keys?: ListKeys;
}

export class List extends GenericInput<string[], ListSettings> {
  public static async prompt(options: string | ListOptions): Promise<string[]> {
    if (typeof options === "string") {
      options = { message: options };
    }

    return new this({
      pointer: blue(Figures.POINTER_SMALL),
      separator: ",",
      minLength: 0,
      maxLength: Infinity,
      minTags: 0,
      maxTags: Infinity,
      ...options,
    }).prompt();
  }

  protected setPrompt(message: string) {
    message += " " + this.settings.pointer + " ";

    const length = new TextEncoder().encode(stripeColors(message)).length;

    const oldInput: string = this.input;
    const oldInputParts: string[] = oldInput.trimLeft().split(this.regexp());

    this.input = oldInputParts.join(`${this.settings.separator} `);

    message += oldInputParts.map((val: string) => underline(val)).join(
      `${this.settings.separator} `,
    );

    const inputDiff = oldInput.length - this.input.length;

    this.index -= inputDiff;

    this.write(message);

    this.screen.cursorTo(length - 1 + this.index);
  }

  protected regexp() {
    return new RegExp(
      this.settings.separator === " " ? ` +` : ` *${this.settings.separator} *`,
    );
  }

  protected addChar(char: string): void {
    switch (char) {
      case this.settings.separator:
        if (
          this.input.length &&
          this.input.trim().slice(-1) !== this.settings.separator
        ) {
          super.addChar(char);
        }
        break;
      default:
        super.addChar(char);
    }
  }

  protected deleteChar(): void {
    if (this.input[this.index - 1] === " ") {
      super.deleteChar();
    }
    super.deleteChar();
  }

  protected validate(value: string): boolean | string {
    if (typeof value !== "string") {
      return false;
    }

    const values = this.transform(value).filter((val) => val !== "");

    for (const val of values) {
      if (val.length < this.settings.minLength) {
        return `Value must be longer then ${this.settings.minLength} but has a length of ${val.length}.`;
      }
      if (val.length > this.settings.maxLength) {
        return `Value can't be longer then ${this.settings.maxLength} but has a length of ${val.length}.`;
      }
    }

    if (values.length < this.settings.minTags) {
      return `The minimum number of tags is ${this.settings.minTags} but got ${values.length}.`;
    }
    if (values.length > this.settings.maxTags) {
      return `The maximum number of tags is ${this.settings.maxTags} but got ${values.length}.`;
    }

    return true;
  }

  protected transform(value: string): string[] {
    return value.trim().split(this.regexp());
  }

  protected format(value: string[]): string {
    return value.join(`, `);
  }
}
