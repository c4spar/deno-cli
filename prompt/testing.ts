import { eraseDown } from "../ansi/ansi_escapes.ts";
import { basename } from "./deps.ts";
import { assertSnapshot } from "https://deno.land/std@0.170.0/testing/snapshot.ts";

export interface AssertPromptSnapshotOptions {
  meta: ImportMeta;
  tests: Record<string, Array<string>>;
  osSuffix?: Array<typeof Deno.build.os>;
  fn(): Promise<unknown>;
}

const encoder = new TextEncoder();

/** Run prompt snapshot tests. */
export function assertPromptSnapshot(
  options: AssertPromptSnapshotOptions,
): Promise<void>;

/** Run prompt snapshot tests. */
export function assertPromptSnapshot(
  meta: ImportMeta,
  tests: Record<string, Array<string>>,
  fn: () => Promise<unknown>,
): Promise<void>;

/**
 * Run prompt snapshot tests.
 *
 * ```ts
 * import { assertPromptSnapshot } from "./testing.ts";
 * import { Input } from "./input.ts";
 *
 * await assertPromptSnapshot({
 *   meta: import.meta,
 *   tests: {
 *     "should enter som text": ["foo bar", "\n"],
 *   },
 *   async fn() {
 *     await Input.prompt({
 *       message: "Whats your name?",
 *       default: "foo",
 *     });
 *   },
 * });
 * ```
 *
 * @param optionsOrMeta import.meta
 * @param tests         Object of tests. Key is the test name and the value is
 *                      an array of input sequences/characters.
 * @param fn            Test function.
 */
export async function assertPromptSnapshot(
  optionsOrMeta: AssertPromptSnapshotOptions | ImportMeta,
  tests?: Record<string, Array<string>>,
  fn?: () => Promise<unknown>,
): Promise<void> {
  const options: AssertPromptSnapshotOptions = "meta" in optionsOrMeta
    ? optionsOrMeta
    : {
      meta: optionsOrMeta,
      tests: tests!,
      fn: fn!,
    };

  if (options.meta.main) {
    await options.fn();
  } else {
    registerTest(options);
  }
}

function registerTest(options: AssertPromptSnapshotOptions) {
  const fileName = basename(options.meta.url);

  Deno.test({
    name: `assert prompt snapshot - ${fileName}`,
    async fn(ctx) {
      const tests = Object.entries(options.tests ?? {});
      if (!tests.length) {
        throw new Error(`No tests defined for: ${options.meta.url}`);
      }

      for (const [name, inputs] of tests) {
        await ctx.step({
          name,
          async fn() {
            const output: string = await runPrompt(options.meta.url, inputs);
            const suffix = options.osSuffix?.includes(Deno.build.os)
              ? `.${Deno.build.os}`
              : "";

            await assertSnapshot(ctx, output, {
              path: `__snapshots__/${fileName}${suffix}.snap`,
            });
          },
        });
      }
    },
  });
}

async function runPrompt(
  url: string,
  inputs: Array<string>,
): Promise<string> {
  const cmd = new Deno.Command("deno", {
    stdin: "piped",
    stdout: "piped",
    stderr: "piped",
    args: [
      "run",
      "--allow-all",
      url,
    ],
    env: {
      NO_COLOR: "true",
    },
  });
  const child: Deno.ChildProcess = cmd.spawn();
  const writer = child.stdin.getWriter();

  for (const input of inputs) {
    await writer.write(encoder.encode(input));
    // Ensure all inputs are processed and rendered separately.
    await new Promise((resolve) => setTimeout(resolve, 600));
  }

  const { success, stdout, stderr } = await child.output();
  writer.releaseLock();
  await child.stdin.close();

  if (!success) {
    throw new Error(`test failed: ${url}`);
  }

  // Add a line break after each test input.
  return "stdout:\n" + new TextDecoder().decode(stdout).replaceAll(
    eraseDown(),
    eraseDown() + "\n",
  ) + "\nstderr:\n" + new TextDecoder().decode(stderr).replaceAll(
    eraseDown(),
    eraseDown() + "\n",
  );
}
