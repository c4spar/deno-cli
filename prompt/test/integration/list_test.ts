import { List } from "../../list.ts";
import { assertPromptSnapshot } from "../../testing.ts";

await assertPromptSnapshot({
  name: "list prompt",
  meta: import.meta,
  osSuffix: ["windows"],
  steps: {
    "should enter some keywords": ["foo,bar", "\n"],
  },
  async fn() {
    await List.prompt({
      message: "Enter some keywords",
      default: ["foo", "far"],
      hint: "some hint...",
    });
  },
});
