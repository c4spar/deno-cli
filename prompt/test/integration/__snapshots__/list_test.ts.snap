export const snapshot = {};

snapshot[`list prompt > should enter some keywords 1`] = `
stdout:
--------------------------
? Enter some keywords (foo, far) ‚Ä∫ 
‚ùØ some hint...[1A[36G[G[0J
? Enter some keywords (foo, far) ‚Ä∫ foo, bar
‚ùØ some hint...[1A[44G[G[0J
? Enter some keywords (foo, far) ‚Ä∫ foo, bar
[?25h[?25h
--------------------------
stderr:
--------------------------
--------------------------
`;
